### ISSUE TITLE
make sure a match between logits and logits_dtype in evaluation_loop

### ISSUE THREAD

**Issue by IT-Forrest** [huggingface/optimum-habana#754]
### System Info

```shell
Optimum Habana version: v.10-release, SynapseAI version: doesn't matter, Docker image: PT 1.14.0:latest
```

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

To reproduce the issue, create a docker container upon a Gaudi device, e.g. Gaudi2
`docker run -it --runtime=habana --name="logits_dtype_err" -v $HOME/tmp:/root/home -w /root/home -e HABANA_VISIBLE_DEVICES=all -e OMPI_MCA_btl_vader_single_copy_mechanism=none --cap-add=sys_ptrace --net=host vault.habana.ai/gaudi-docker/1.14.0/ubuntu22.04/habanalabs/pytorch-installer-2.1.1:latest`

Then access the docker container `logits_dtype_err` and run the following scripts to reproduce the err with 2 HPU cards:
```
python3 -m pip install git+https://github.com/HabanaAI/DeepSpeed.git@1.14.0;
python3 -m pip install git+https://github.com/huggingface/optimum-habana.git@v1.10-release;
git clone https://github.com/huggingface/optimum-habana.git
cd optimum-habana/examples/language-modeling
python ../gaudi_spawn.py  --world_size 2 --use_deepspeed run_clm.py --model_name_or_path EleutherAI/gpt-neox-20b --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --per_device_eval_batch_size 2 --do_eval --output_dir ./gpt-neox-20b --gaudi_config_name Habana/gpt2 --use_habana --use_lazy_mode --use_hpu_graphs_for_inference --overwrite_output_dir
```

To supplement, this issue is also ever reported in another issue https://github.com/huggingface/optimum-habana/pull/457 the model in that issue is **chatglm2**, which is different from the current model, **GPT-Neox**

### Expected behavior

By the end of the evaluation loop, it will hit the **TypeError: Field elements must be 2- or 3-tuples, got ''float32''**
The stack trace is as follows:
![image](https://github.com/huggingface/optimum-habana/assets/9160073/24bf498a-7527-4d32-8828-7ad394238eec)

PS: we expect the evaluation could be accomplished successfully as follows:
```
***** eval metrics *****
  eval_accuracy               =     0.4935
  eval_loss                   =     2.3582
  eval_runtime                = 0:00:35.53
  eval_samples                =        241
  eval_samples_per_second     =      6.781
  eval_steps_per_second       =      1.716
  max_memory_allocated (GB)   =      91.27
  memory_allocated (GB)       =      90.86
  perplexity                  =    10.5718
  total_memory_available (GB) =      94.62
[INFO|modelcard.py:452] 2024-02-23 15:23:51,389 >> Dropping the following result as it does not have all the necessary fields:
{'task': {'name': 'Causal Language Modeling', 'type': 'text-generation'}, 'dataset': {'name': 'wikitext wikitext-2-raw-v1', 'type': 'wikitext', 'args': 'wikitext-2-raw-v1'}}
[2024-02-23 15:24:00,229] [INFO] [launch.py:348:main] Process 346467 exits successfully.
[2024-02-23 15:24:01,231] [INFO] [launch.py:348:main] Process 346466 exits successfully.
```

**Comment by IT-Forrest** [comment]
Hi @regisss, this issue is similar to the one in https://github.com/huggingface/optimum-habana/pull/457
The difference is that the model in that issue is chatglm2, while it is GPT-Neox for my situation.

The reproduction steps have been provided, and the solution can be referred to https://github.com/huggingface/optimum-habana/pull/457/files
Kindly let me know if you need further info from my side, thanks

**Comment by regisss** [comment]
@IT-Forrest Can you try with the branch of [this PR](https://github.com/huggingface/optimum-habana/pull/769) and let me know if that works for you too please?

The fix proposed in #457 works for this example but may lead to other issues with other use cases.

**Comment by IT-Forrest** [comment]
Thanks for your feedback Regisss. I have tried the PR that you suggested https://github.com/huggingface/optimum-habana/pull/769, and it could resolve the `TypeError: Field elements must be 2- or 3-tuples, got ''float32''` Please move ahead to merge the PR.
